### ISSUE TITLE
MnistExperiment crashes: bias gradient dimension mismatch on first training step

### ISSUE THREAD

**Issue by LeoBuron** [es-ude/OnDeviceTraining#69]
## Bug

`MnistExperiment` crashes immediately on the first training step with:

```
[ARITHMETIC: doDimensionsMatch] Dim 1: 0, Dims 2: 20
[ARITHMETIC: floatPointWiseArithmeticInplace] Dimensions don't match
```

## Reproduction

```bash
cmake --preset unit_test_debug
cmake --build --preset unit_test_debug --target MnistExperiment
cd build/unit_test_debug/experiments && ./MnistExperiment
```

Requires MNIST `.npy` files generated via `uv run test/unit/data_loader/MNISTLoader.py`.

## Analysis

The error occurs in `floatPointWiseArithmeticInplace` (bias gradient accumulation during the backward pass). A gradient tensor has a first dimension of 0 where 20 is expected — the bias grad shape for the first linear layer `[1, 20]` is being compared against a tensor with shape containing 0.

This is a **pre-existing bug** — it reproduces on `main` (commit `152e95a`) before the #66 training_loop refactoring. Likely introduced or exposed by the #67 memory leak fixes, or was always latent but masked by a different code path.

## Notes

- The unit tests for `calculateGrads` / `calculateGradsSequential` pass — they use a small 1-layer model with MSE loss, not the 4-layer MNIST model with CROSS_ENTROPY + Softmax
- The dimension `20` matches the first hidden layer size in `MnistExperiment.c` (784 → **20** → 10)

**Comment by LeoBuron** [comment]
Fixed in #73.
